src/main/java/ch/njol/skript/effects/EffToggle.java

```diff
@@ -1,84 +1,181 @@
+	private Expression<?> togglables;
+	private Action action;
+	private Type type;
+
 	@Override
-	public boolean init(final Expression<?>[] vars, final int matchedPattern, final Kleenean isDelayed, final ParseResult parseResult) {
-		blocks = (Expression<Block>) vars[0];
-		toggle = matchedPattern - 1;
+	public boolean init(Expression<?>[] expressions, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		togglables = expressions[0];
+		action = patterns.getInfo(matchedPattern);
+
+		// Determine expression type using the enum method
+		type = Type.fromClass(togglables.getReturnType());
+		
+		// Validate based on type
+		if (type == Type.BOOLEANS && 
+			!Changer.ChangerUtils.acceptsChange(togglables, Changer.ChangeMode.SET, Boolean.class)) {
+			Skript.error("Cannot toggle '" + togglables + "' as it cannot be set to booleans.");
+			return false;
+		} else if (type == Type.MIXED && !Changer.ChangerUtils.acceptsChange(togglables, Changer.ChangeMode.SET, Object.class)) {
+			Skript.error("Cannot toggle '" + togglables + "' as it cannot be set to both blocks and booleans.");
+			return false;
+		}
+
 		return true;
 	}
 
 	@Override
-	protected void execute(final Event e) {
-		for (Block b : blocks.getArray(e)) {
-			BlockData data = b.getBlockData();
-			if (toggle == -1) {
-				if (data instanceof Openable)
-					((Openable) data).setOpen(false);
-				else if (data instanceof Powerable)
-					((Powerable) data).setPowered(false);
-			} else if (toggle == 1) {
-				if (data instanceof Openable)
-					((Openable) data).setOpen(true);
-				else if (data instanceof Powerable)
-					((Powerable) data).setPowered(true);
-			} else {
-				if (data instanceof Openable) // open = NOT was open
-					((Openable) data).setOpen(!((Openable) data).isOpen());
-				else if (data instanceof Powerable) // power = NOT power
-					((Powerable) data).setPowered(!((Powerable) data).isPowered());
+	protected void execute(Event event) {
+		switch (type) {
+			case BOOLEANS -> toggleBooleans(event);
+			case BLOCKS -> toggleBlocks(event);
+			case MIXED -> toggleMixed(event);
+		}
+	}
+
+	/**
+	 * Toggles blocks by opening/closing or powering/unpowering them.
+	 * @param event the event used for evaluation
+	 */
+	private void toggleBlocks(Event event) {
+		for (Object obj : togglables.getArray(event)) {
+			if (obj instanceof Block block) {
+				toggleSingleBlock(block);
 			}
-			
-			b.setBlockData(data);
 		}
 	}
 
+	/**
+	 * Toggles a single block, either by opening/closing or powering/unpowering it.
+	 * @param block The block to toggle
+	 */
+	private void toggleSingleBlock(@NotNull Block block) {
+		BlockData data = block.getBlockData();
+		if (data instanceof Openable openable) {
+			openable.setOpen(action.apply(openable.isOpen()));
+			block.setBlockData(data);
+		} else if (data instanceof Powerable powerable) {
+			powerable.setPowered(action.apply(powerable.isPowered()));
+			block.setBlockData(data);
+		}
```

Comment:
```suggestion
		BlockData data = block.getBlockData();
		if (data instanceof Openable openable) {
			openable.setOpen(action.apply(openable.isOpen()));
		} else if (data instanceof Powerable powerable) {
			powerable.setPowered(action.apply(powerable.isPowered()));
		}
		block.setBlockData(data);
```
We can pull out the common portion of updating the block data on the block